core/shared/src/main/scala/spire/algebra/Gcd.scala

```diff
@@ -0,0 +1,25 @@
+package spire.algebra
+
+import annotation.tailrec
+import scala.{specialized => sp}
+
+import spire.syntax.eq._
+import spire.syntax.euclideanRing._
+
+trait Gcd[@sp(Int, Long, Float, Double) A] extends Any {
+  def gcd(a: A, b: A): A
+  def lcm(a: A, b: A): A
+
+  @tailrec final def euclid(a: A, b: A)(implicit eq: Eq[A], er: EuclideanRing[A]): A =
+    if (b === er.zero) a else euclid(b, a % b)
+}
+
+object Gcd {
+  @inline final def apply[A](implicit ev: Gcd[A]): Gcd[A] = ev
+
+  implicit def fromEuclideanRing[@sp(Int, Long, Float, Double) A: EuclideanRing: Eq] =
```

Comment:
Also - really appreciate you taking a look at the implementations. In particular, the implementations for `Rational`, `Float`, and `Double` are somewhat involved. If the `Polynomial` one can be done w/ just euclid (perhaps w/ some adjustments to `euclid`), that'd be great. I'm not super clear on why the special case is necessary :\